### PR TITLE
Add manual triage view

### PR DESCRIPTION
**Pull request by alejandrosame**
This view allows users to filter CVEs and packages, link them and create NixpkgsIssue from their selections.

Model changes:

- Added **SearchVectorField** to improve performance of full-text search.
- Indices added for performance improvements:
  - **GinIndex** for SearchVectorFields.
  - **BTreeIndex** for fields to be searched for filtering and aggregation.
  
 Part of https://github.com/Nix-Security-WG/nix-security-tracker/issues/5

### CODE REVIEW

File: src/website/webview/views.py [hunk 1]
```diff
@@ -1,16 +1,363 @@
...
 class HomeView(TemplateView):
     template_name = "home_view.html"
 
 
+@method_decorator(login_required, name="dispatch")
+class TriageView(TemplateView):
+    template_name = "triage_view.html"
+    paginate_by = 10
+    pages_on_each_side = 2
+    pages_on_ends = 1
```
Comment (fricklerhandwerk) [inline src/website/webview/views.py:51]:
```suggestion
    # parameters for Paginator 
    paginate_by = 10
    pages_on_each_side = 2
    pages_on_ends = 1
```

File: src/website/webview/views.py [hunk 2]
```diff
@@ -1,16 +1,363 @@
...
+
+    def _get_pkg_page(
+        self, search_pkgs: str | None, pkg_page_number: int
+    ) -> tuple[Page, CustomCountPaginator]:
+        pkg_qs, pkg_count_function = self._get_pkg_qs(search_pkgs)
+        pkg_paginator = CustomCountPaginator(
+            pkg_qs, self.paginate_by, custom_count=pkg_count_function
+        )
+        pkg_page = pkg_paginator.get_page(pkg_page_number)
+        return pkg_page, pkg_paginator
```
Comment (fricklerhandwerk) [inline src/website/webview/views.py:293]: Not critical: One could pass what's currently `_get_cve_qs` and `_get_pkgs_qs` as a parameter `_get_queryset`, and rename the search string and page number variables to be more generic. But it's okay for now, the duplication is not super confusing.

File: src/website/webview/views.py [hunk 3]
```diff
@@ -1,16 +1,363 @@
...
+            return cve_qs, cve_count_function
+
+    def _get_pkg_qs(
+        self, search_pkgs: str | None
+    ) -> tuple["ValuesQuerySet[NixDerivation, dict[str, Any]]", Callable[[], int]]:
+        if search_pkgs:
+            # Do a 2-rank search to prevent description contents from penalizing hits on "name" and "attribute"
+            search_query = SearchQuery(search_pkgs)
+            # Check https://www.postgresql.org/docs/current/textsearch-controls.html#TEXTSEARCH-RANKING
+            # for the meaning of normalization values.
```
Comment (fricklerhandwerk) [inline src/website/webview/views.py:207]: Copy that comment to the other function, so a cursory reader will not miss it

**Review by fricklerhandwerk — APPROVED** [review]
There are some tiny cosmetic things left, such as check boxes being invalidated on navigating between pages. The JS to handle that is ugly, but okay for now since we're not taking too much care for the front-end side of things.

File: src/website/webview/views.py (hunk 1 above)
Comment (alejandrosame) [reply src/website/webview/views.py:52]: Applied

File: src/website/webview/views.py (hunk 3 above)
Comment (alejandrosame) [reply src/website/webview/views.py:213]: Applied

File: src/website/webview/views.py [hunk 4]
```diff
@@ -1,16 +1,363 @@
...
+                    "-max_rank_affected",
+                    "-max_rank_cpes",
+                    "-max_rank_descriptions",
+                    "id",
+                )
+            )
+
+            # To efficiently count is best to filter all relevant m2m related fields
+            # disregarding their order. Otherwise, the count will and paginator will
+            # force grouping (which is costly in this case) twice.
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply src/website/webview/views.py]: Applied

File: src/website/webview/paginators.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+from typing import Any
+
+from django.core.paginator import Paginator
+from django.utils.functional import cached_property
+
+
+class CustomCountPaginator(Paginator):
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply src/website/webview/paginators.py:13]: Applied

File: src/website/webview/views.py [hunk 6]
```diff
@@ -1,16 +1,363 @@
...
+            on_ends=self.pages_on_ends,
+        )
+
+        context["search_cves"] = search_cves
+        context["search_pkgs"] = search_pkgs
+
+        return context
+
+    def get(self, request: HttpRequest, *args: Any, **kwargs: Any) -> HttpResponse:
+        context = self.get_context_data(**kwargs)
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply src/website/webview/views.py:340]: Applied

File: src/website/webview/views.py [hunk 7]
```diff
@@ -1,16 +1,363 @@
...
+        search_cves = self.request.GET.get("search_cves")
+        search_pkgs = self.request.GET.get("search_pkgs")
+        cve_page_number = int(self.request.GET.get("cve_page", 1))
+        pkg_page_number = int(self.request.GET.get("pkg_page", 1))
+
+        cve_page, cve_paginator = self._get_cve_page(search_cves, cve_page_number)
+        pkg_page, pkg_paginator = self._get_pkg_page(search_pkgs, pkg_page_number)
+
+        context["cve_list"] = cve_page
+        context["pkg_list"] = pkg_page
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply src/website/webview/views.py]: Applied

**Comment by fricklerhandwerk** [comment]
Please rewrite history and I'll merge.

File: src/website/webview/templates/triage_view.html [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+<script>
+    {% comment %}
+        NOTE(alejandrosame): If we want to allow selecting packages and CVEs from different pages,
+        then we'll have to refactor to manage state with localstorage.
+    {% endcomment %}
+    // On refresh, set all checkboxes as unchecked
+    const allCheckboxes = document.querySelectorAll('input[type="checkbox"]')
+    allCheckboxes.forEach(function (checkbox) {
+        checkbox.checked = false;
+    });
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply src/website/webview/templates/triage_view.html]: This should be fixed now that there's some front-end state management for the checkboxes.

**Comment by alejandrosame** [comment]
> Please rewrite history and I'll merge.

Doing it after this last commit. Now there's some basic checkbox state management, but I'm skipping the CSS/HTML change suggestions

**Pull request merged**